### PR TITLE
Allocation improvements

### PR DESCRIPTION
**Pull request by isaachier**

### CODE REVIEW

File: src/apps/benchmarks/benchmarkPolyfill.c [hunk 1]
```diff
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#include <string.h>
```
Comment (dfellis) [inline src/apps/benchmarks/benchmarkPolyfill.c]: I don't see any code in here using `string.h`

File: src/apps/filters/hexRange.c [hunk 2]
```diff
@@ -30,13 +30,15 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
```
Comment (dfellis) [inline src/apps/filters/hexRange.c]: Same here. `printf` is in `stdio`, I don't see why `string` was included.

File: src/apps/filters/kRing.c [hunk 3]
```diff
@@ -25,22 +25,22 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
```
Comment (dfellis) [inline src/apps/filters/kRing.c]: Please remove this.

File: src/apps/miscapps/generateBaseCellNeighbors.c [hunk 4]
```diff
@@ -237,18 +226,19 @@ void generate() {
 
     printf("const int baseCellNeighbors[NUM_BASE_CELLS][7] = {\n");
     for (int i = 0; i < NUM_BASE_CELLS; i++) {
-        char* neighborStrings[7] = {0, 0, 0, 0, 0, 0, 0};
+        printf("{");
```
Comment (dfellis) [inline src/apps/miscapps/generateBaseCellNeighbors.c]: Please undo the changes you made here. This approach will print garbage around the `INVALID_BASE_CELL` printout, instead of only printing that message when reaching that condition, so it will not generate the base cell neighbors code correctly, anymore.

File: src/h3lib/lib/algos.c [hunk 5]
```diff
@@ -22,6 +22,7 @@
 #include <math.h>
 #include <stdbool.h>
 #include <stdlib.h>
+#include <string.h>
```
Comment (dfellis) [inline src/h3lib/lib/algos.c]: Remove this.

File: src/apps/benchmarks/benchmarkPolyfill.c (hunk 1 above)
Comment (isaachier) [reply src/apps/benchmarks/benchmarkPolyfill.c]: `memset` for all of these comments.

File: src/apps/miscapps/generateBaseCellNeighbors.c (hunk 4 above)
Comment (isaachier) [reply src/apps/miscapps/generateBaseCellNeighbors.c]: Now I'm very confused. The difference between the old code and this is the removal of the `generateBaseCellOutput`, which simply formatted strings that are printed here. Instead, I used identical code here to write the digits/macro directly to stdout.

File: src/h3lib/lib/algos.c [hunk 6]
```diff
@@ -90,9 +91,8 @@ int H3_EXPORT(maxKringSize)(int k) {
  */
 void H3_EXPORT(kRing)(H3Index origin, int k, H3Index* out) {
     int maxIdx = H3_EXPORT(maxKringSize)(k);
-    int* distances = calloc(maxIdx, sizeof(int));
+    int distances[maxIdx];
```
Comment (isaacbrodsky) [inline src/h3lib/lib/algos.c]: I was a little concerned this does not initialize `distances`. It might be that `distances` cannot be used before initializing. Not sure if you want to add a memset here for clarity.

File: src/h3lib/lib/h3Index.c [hunk 7]
```diff
@@ -340,17 +337,11 @@ int H3_EXPORT(compact)(const H3Index* h3Set, H3Index* compactedSet,
...
+        memset(hashSetArray, 0, sizeof(hashSetArray));
         compactedSetOffset += uncompactableCount;
-        remainingHexes = compactableHexes;
+        memcpy(remainingHexes, compactableHexes, sizeof(compactableHexes));
         numRemainingHexes = compactableCount;
     }
-    // Final cleanup
-    free(remainingHexes);
-    free(hashSetArray);
     return 0;
```
Comment (isaacbrodsky) [inline src/h3lib/lib/h3Index.c]: It looks like this line is no longer reached in tests, perhaps because of the change around line 276. Could you change to preserve 100% coverage?

File: src/apps/miscapps/generateBaseCellNeighbors.c (hunk 4 above)
Comment (isaacbrodsky) [reply src/apps/miscapps/generateBaseCellNeighbors.c]: @dfellis Could you paste the output you are seeing or expecting? I ran this branch and got the same output as before.

Comment (dfellis) [reply src/apps/miscapps/generateBaseCellNeighbors.c]: My mistake on the reading of his changes. I did not run his branch.

File: src/h3lib/lib/algos.c (hunk 6 above)
Comment (isaachier) [reply src/h3lib/lib/algos.c]: Good catch. Was almost certain I followed them all with a `memset`.

Comment (isaachier) [reply src/h3lib/lib/algos.c]: Fixed.

File: src/h3lib/lib/algos.c [hunk 8]
```diff
@@ -741,7 +742,8 @@ void H3_EXPORT(polyfill)(const GeoPolygon* geoPolygon, int res, H3Index* out) {
     // This first part is identical to the maxPolyfillSize above.
 
     // Get the bounding boxes for the polygon and any holes
-    BBox* bboxes = calloc(sizeof(BBox), geoPolygon->numHoles + 1);
+    BBox bboxes[geoPolygon->numHoles + 1];
+    memset(bboxes, 0, sizeof(BBox));
```
Comment (isaachier) [inline src/h3lib/lib/algos.c]: This is a mistake. Let me fix now.

File: src/h3lib/lib/algos.c [hunk 9]
```diff
@@ -741,7 +742,8 @@ void H3_EXPORT(polyfill)(const GeoPolygon* geoPolygon, int res, H3Index* out) {
     // This first part is identical to the maxPolyfillSize above.
 
     // Get the bounding boxes for the polygon and any holes
-    BBox* bboxes = calloc(sizeof(BBox), geoPolygon->numHoles + 1);
+    BBox bboxes[geoPolygon->numHoles + 1];
+    memset(bboxes, 0, sizeof(bboxes));
```
Comment (isaachier) [inline src/h3lib/lib/algos.c]: Fixed.

File: src/h3lib/lib/h3Index.c (hunk 7 above)
Comment (isaachier) [reply src/h3lib/lib/h3Index.c]: Sure. Will push fix for that now.

**Comment by isaachier** [comment]
Think I addressed all the comments.

**Comment by isaacbrodsky** [comment]
I tested this on MSVC (Visual Studio 2017) and it seems it does not support variable length arrays. I would prefer the repo is limited to the subset of C99/C11 that MSVC supports to maximize how broadly it can be used.

File: src/h3lib/lib/linkedGeo.c [hunk 10]
```diff
@@ -103,35 +106,27 @@ LinkedGeoCoord* addLinkedCoord(LinkedGeoLoop* loop, const GeoCoord* vertex) {
...
-    LinkedGeoCoord* nextCoord;
     // flag to skip the input polygon
-    int skip = 1;
-    while (currentPolygon != NULL) {
-        currentLoop = currentPolygon->first;
-        while (currentLoop != NULL) {
-            currentCoord = currentLoop->first;
-            while (currentCoord != NULL) {
+    bool skip = true;
+    for (LinkedGeoPolygon *currentPolygon = polygon, *nextPolygon;
```
Comment (nrabinowitz) [inline src/h3lib/lib/linkedGeo.c]: Nit: I believe our style in this project is `LinkedGeoPolygon* currentPolygon`, can we use that here? Or would that require `LinkedGeoPolygon* currentPolygon, LinkedGeoPolygon* nextPolygon`?

Comment (isaachier) [reply src/h3lib/lib/linkedGeo.c]: You mean the pointer alignment? Ya it's a problem. `clang-format` isn't even fixing it. The only alternative is `LinkedGeoPolygon* nextPolygon;` followed by `for (LinkedGeoPolygon* currentPolygon = ...)`. I once answered a question about this on StackOverflow here: https://stackoverflow.com/a/16490876/1930331.

File: cmake/CheckAlloca.cmake [hunk 11]
```diff
@@ -0,0 +1,21 @@
+if(__check_alloca)
```
Comment (isaacbrodsky) [inline cmake/CheckAlloca.cmake]: Please include the license header in all source files.

File: src/h3lib/include/stackAlloc.h [hunk 12]
```diff
@@ -0,0 +1,44 @@
+#ifndef STACKALLOC_H
```
Comment (isaacbrodsky) [inline src/h3lib/include/stackAlloc.h]: Please also include a `@file`/`@brief` comment describing the purpose of this file.

File: src/apps/miscapps/generateBaseCellNeighbors.c [hunk 13]
```diff
@@ -237,24 +223,25 @@ void generate() {
...
+                printf("%d", baseCellNeighbors[i][j]);
+            } else {
+                printf("INVALID_BASE_CELL");
+            }
         }
-        printf("{%s, %s, %s, %s, %s, %s, %s}, // base cell %d%s\n",
-               neighborStrings[0], neighborStrings[1], neighborStrings[2],
-               neighborStrings[3], neighborStrings[4], neighborStrings[5],
-               neighborStrings[6], i,
+        printf("}, /* base cell %d%s */\n", i,
```
Comment (isaacbrodsky) [inline src/apps/miscapps/generateBaseCellNeighbors.c]: Since the formatting of the comments changes, could you rerun generateBaseCellNeighbors and update baseCells.c?

**Comment by isaacbrodsky** [comment]
I successfully built and tested this on MSVC :) - #34 will make that automated.

File: src/apps/miscapps/generateBaseCellNeighbors.c (hunk 13 above)
Comment (isaachier) [reply src/apps/miscapps/generateBaseCellNeighbors.c]: Sure thing.

File: cmake/CheckAlloca.cmake (hunk 11 above)
Comment (isaachier) [reply cmake/CheckAlloca.cmake]: OK.

File: src/h3lib/include/stackAlloc.h (hunk 12 above)
Comment (isaachier) [reply src/h3lib/include/stackAlloc.h]: Good idea

**Comment by isaachier** [comment]
Luckily even MSVC has the alloca function. It's a bit ugly but this should definitely allow some form of VLA on all popular platforms.

File: src/apps/miscapps/generateBaseCellNeighbors.c [hunk 14]
```diff
@@ -221,9 +243,52 @@ static void generate() {
 
     auditBaseCellNeighbors(baseCellNeighbors, baseCellRotations);
 
+    time_t rawTime;
+    time(&rawTime);
+    struct tm tmStruct;
+    reentrant_gmtime(&tmStruct, &rawTime);
+    const char licenseFormat[] =
+        "/*\n"
+        " * Copyright 2016-%d Uber Technologies, Inc.\n"
```
Comment (isaacbrodsky) [inline src/apps/miscapps/generateBaseCellNeighbors.c]: This seems like a lot of work just to automatically format the date. I don't think the parts of the baseCells.c file unrelated to the baseCellNeighbors/baseCellNeighbor60CCWRots should be generated. (Those two were the only parts of the file that were generated from this - maybe those two could be split off to their own, generated file.)

Could you revert this change, and we can revisit it in another issue or PR?
